### PR TITLE
feat: auto respect rbac for discovery/sync

### PR DESCRIPTION
**Pull request by gdsoumya**
This PR enables controller to automatically stop watching for resources that it does not have permission to access without having to manually set resource exclusions.

Depends on https://github.com/argoproj/gitops-engine/pull/532

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)). 
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.

Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -295,3 +295,5 @@ replace (
 	k8s.io/pod-security-admission => k8s.io/pod-security-admission v0.24.2
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.24.2
 )
+
+replace github.com/argoproj/gitops-engine => github.com/gdsoumya/gitops-engine v0.0.0-20230706170353-56c89ca1b7cd
```
Comment (gdsoumya) [inline go.mod:299]: needs to be updated once https://github.com/argoproj/gitops-engine/pull/532 is merged

**Review by alexmt — APPROVED** [review]
LGTM!

**Comment by jannfis** [comment]
The end-to-end tests seem to fail at always the same test, so I suppose it's a real bug, not flakyness.

Can you PTAL?

**Comment by gdsoumya** [comment]
@jannfis on it.

File: go.mod [hunk 2]
```diff
@@ -326,3 +326,5 @@ replace (
 	k8s.io/pod-security-admission => k8s.io/pod-security-admission v0.24.2
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.24.2
 )
+
+replace github.com/argoproj/gitops-engine => github.com/gdsoumya/gitops-engine v0.0.0-20230906104903-608e459b2a99
```
Comment (gdsoumya) [inline go.mod:330]: Needs to be updated before merge, depends on https://github.com/argoproj/gitops-engine/pull/539

**Review by alexmt — APPROVED** [review]

**Pull request merged**